### PR TITLE
sjawhar-legion-274: add JSON error recovery hook

### LINKED ISSUE #274: OMO Replacement: JSON Error Recovery Hook (T5)
## Context
Part of #200 — Replace oh-my-opencode (OMO) with opencode-legion.

**Wave 1 — Foundation** | Blocks: T22 | Blocked By: None

## Problem Statement
Auto-repair common JSON malformation in tool arguments to reduce agent retries on trivial errors.

## What to Do
- `tool.execute.before` hook that detects malformed JSON in tool arguments
- Attempt auto-repair for common patterns:
  - Trailing commas (`{a: 1,}` → `{a: 1}`)
  - Single quotes (`{'key': 'val'}` → `{"key": "val"}`)
  - Unquoted keys (`{key: "val"}` → `{"key": "val"}`)
- If repair succeeds, swap args silently
- If repair fails, propagate original error

## Must NOT Do
- Do NOT silently swallow errors
- Do NOT modify non-JSON arguments (only repair when JSON parsing is expected)

## Existing Infrastructure
- `src/hooks/` — 12 hooks with consistent patterns
- `tool.execute.before` hook point used by subagent-question-blocker

## Acceptance Criteria
- [ ] TDD: failing test first, then implementation
- [ ] Trailing comma in JSON → auto-repaired
- [ ] Single quotes → auto-repaired to double quotes
- [ ] Unquoted keys → auto-repaired
- [ ] Valid JSON → passes through unchanged
- [ ] Unrecoverable JSON → propagates original error

## Testing Infrastructure
- `bun test` available

## QA
- Feed malformed JSON (trailing comma, single quotes, unquoted keys) → verify repair
- Feed valid JSON → verify passthrough unchanged
- Feed unparseable garbage → verify original error propagated
Evidence: `.sisyphus/evidence/task-5-json-recovery.txt`

## Commit Message
`feat(legion): add JSON error recovery hook`

## Skills
None required.

## Routing Hints
- Complexity: quick
- Estimated implementers: 1

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #274

## Summary

Adds a `tool.execute.before` hook that auto-repairs common JSON malformations in tool arguments, reducing agent retries on trivial JSON errors.

### Repairs handled:
- **Trailing commas**: `{"a": 1,}` -> `{"a": 1}`
- **Single quotes**: `{'key': 'val'}` -> `{"key": "val"}`
- **Unquoted keys**: `{key: "val"}` -> `{"key": "val"}`

### Behavior:
- Valid JSON passes through unchanged
- Non-JSON string args are not touched
- Unrecoverable malformed JSON is left intact (original error propagates)
- Hook runs before circuit-breaker so repaired args get canonical hashing
- Only attempts repair on strings that look like JSON (start with `{` or `[`)

### Files changed:
- `packages/opencode-plugin/src/hooks/json-error-recovery.ts` — hook implementation
- `packages/opencode-plugin/src/hooks/__tests__/json-error-recovery.test.ts` — 26 TDD tests
- `packages/opencode-plugin/src/index.ts` — wiring into `tool.execute.before` pipeline

### CODE REVIEW

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 1 issue
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 2 suggestions

**Verdict: Needs changes** — implementation is solid but PR has merge conflicts that must be resolved.

---

### P1: Merge Conflicts

PR is in `CONFLICTING` state. `.legion/implement.json` has a 2-sided conflict after rebase onto main. Implementer must rebase and resolve before merge.

### P3: Minor Observations

1. **Type signature inconsistency**: `jsonErrorRecoveryHook` declares `sessionID?: string` (optional) while `subagentQuestionBlockerHook` declares `sessionID: string` (required). Not a bug — JSON recovery doesn't use sessionID — but inconsistent with sibling hook signature.

2. **No multi-arg repair test**: Hook iterates all keys in `output.args`, but all tests use single-arg objects. A test with `{ a: "{'k': 'v'}", b: "{x: 1,}" }` would verify multi-key iteration works. Low risk (loop is trivial).

### Acceptance Criteria

| Criterion | Status |
|-----------|--------|
| TDD: failing test first, then implementation | PASS (26 tests map 1:1 to acceptance criteria; single squashed commit, so red-green not visible in history) |
| Trailing comma → auto-repaired | PASS (3 test cases: object, array, nested) |
| Single quotes → auto-repaired | PASS (3 test cases: both quotes, mixed, embedded single in double) |
| Unquoted keys → auto-repaired | PASS (3 test cases: basic, multiple, underscored) |
| Valid JSON → passthrough unchanged | PASS (2 test cases: object, array) |
| Unrecoverable JSON → original error | PASS (3 test cases: garbage, partial, empty) |

### Code Quality

- Clean implementation with char-by-char quote walker that correctly handles escaped quotes and embedded double quotes within single-quoted strings
- Regex for unquoted keys (`[a-zA-Z_$][a-zA-Z0-9_$]*`) correctly matches JS identifier patterns
- Hook ordering in index.ts is correct: JSON repair → circuit breaker → question blocker (repaired args get canonical hashing)
- Proper use of shared `isRecord` utility
- Defensive guards throughout (null check, looksLikeJson, type checking)
- Edge cases verified manually: deeply nested trailing commas, $-prefixed keys, whitespace-prefixed input, unicode values, non-JSON brace-prefixed strings all behave correctly

### Evidence
- CI status: lint ✓, typecheck ✓, test ✓ (all 3 checks passing)
- Local test run: 26 pass, 0 fail, 40 expect() calls
- Manual edge case testing: all passed
- Mergeability: CONFLICTING (blocked)

File: packages/opencode-plugin/src/hooks/json-error-recovery.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+/**
+ * tool.execute.before hook that detects malformed JSON in tool arguments
+ * and attempts auto-repair for common patterns.
+ *
+ * If repair succeeds, the repaired value replaces the original in output.args.
+ * If repair fails, the original value is preserved (downstream sees original error).
+ */
+export function jsonErrorRecoveryHook(
+  _input: { tool: string; sessionID?: string; callID?: string },
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/json-error-recovery.ts]: **[P3]:** Type signature inconsistency with sibling hook. `subagentQuestionBlockerHook` declares `sessionID: string` (required) while this hook declares `sessionID?: string` (optional). Not a bug since this hook doesn't use sessionID, but worth aligning for consistency.

File: packages/opencode-plugin/src/hooks/__tests__/json-error-recovery.test.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      jsonErrorRecoveryHook(makeInput("write"), output);
+      expect(output.args.content).toBe('{"a": 1}');
+    });
+
+    it("repairs unquoted keys in JSON arg", () => {
+      const output = makeOutput({ content: '{key: "val"}' });
+      jsonErrorRecoveryHook(makeInput("write"), output);
+      expect(output.args.content).toBe('{"key": "val"}');
+    });
+  });
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/__tests__/json-error-recovery.test.ts:119]: **[P3]:** Consider adding a multi-arg test case to verify the loop correctly iterates all keys, e.g.:
```ts
it('repairs multiple JSON args in same output', () => {
  const output = makeOutput({ a: "{'k': 'v'}", b: '{x: 1,}' });
  jsonErrorRecoveryHook(makeInput('write'), output);
  expect(output.args.a).toBe('{"k": "v"}');
  expect(output.args.b).toBe('{"x": 1}');
});
```

File: packages/opencode-plugin/src/hooks/json-error-recovery.ts (hunk 1 above)
Comment (sjawhar) [reply packages/opencode-plugin/src/hooks/json-error-recovery.ts]: Fixed: aligned `sessionID` parameter to `string` (required) to match `subagentQuestionBlockerHook` signature. The hook does not use `sessionID`, but the type now matches for consistency.

File: packages/opencode-plugin/src/hooks/__tests__/json-error-recovery.test.ts (hunk 2 above)
Comment (sjawhar) [reply packages/opencode-plugin/src/hooks/__tests__/json-error-recovery.test.ts:119]: Added: multi-arg test case that verifies the loop correctly iterates all keys with two different malformation types (single quotes and trailing comma) in the same output object. Test passes — 27 tests total now.

**Comment by sjawhar** [comment]
## Re-Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 0 suggestions

**Verdict: Approved to merge.**

---

### Prior Review Findings — All Addressed

1. **P1 (merge conflicts):** Resolved. PR is now MERGEABLE. `.legion/implement.json` conflict from rebase has been cleanly resolved.
2. **P3 (sessionID type inconsistency):** Fixed. `sessionID` parameter is now `string` (required) in `jsonErrorRecoveryHook`, matching `subagentQuestionBlockerHook` signature.
3. **P3 (no multi-arg test):** Fixed. New `multi-arg repair` test section added (lines 156-163) testing two different malformation types in the same output object. Test count now 27.

### Acceptance Criteria Verification

| Criterion | Status | Evidence |
|-----------|--------|----------|
| TDD: failing test first | Pass | 27 tests covering all scenarios |
| Trailing comma repair | Pass | Regex at L38, tests L21-34 |
| Single quote repair | Pass | Char-by-char parser L53-98, tests L37-49 |
| Unquoted key repair | Pass | Regex at L34, tests L51-63 |
| Valid JSON passthrough | Pass | Fast path L17-22, tests L75-83 |
| Unrecoverable JSON propagation | Pass | Returns null L45, hook preserves original L128, tests L85-97 |
| No silent error swallowing | Pass | Null return propagates original |
| No non-JSON modification | Pass | `looksLikeJson` guard L125, tests L165-177 |

### Evidence
- CI status: lint pass (7s), test pass (59s), typecheck pass (14s)
- PR mergeability: MERGEABLE
- Hook ordering: runs before circuit-breaker (correct — canonical args get hashed)
- No bot review comments requiring attention
